src/vos/vos_pool_scrub.c

```diff
@@ -147,91 +230,108 @@ sc_get_rec_in_chunk_at_idx(const struct scrub_ctx *ctx, uint32_t i)
-			now, ctx->sc_pool->sp_scrub_freq_sec,
-			ctx->sc_pool_last_csum_calcs,
-			/* -1 to convert to index (from count) */
-			ctx->sc_pool_csum_calcs - 1);
+	if (ds_notify_ras_event != NULL) {
+		ds_notify_ras_event(RAS_POOL_CORRUPTION_DETECTED,
+				    "Data corruption detected",
+				    RAS_TYPE_INFO,
+				    RAS_SEV_ERROR, NULL, NULL, NULL, NULL,
+				    &ctx->sc_pool_uuid, sc_cont_uuid(ctx),
+				    NULL, NULL, NULL);
 	}
+}
 
-	if (msec_between == 0)
-		sc_yield(ctx);
-	else
-		sc_sleep(ctx, msec_between);
-
-	sc_credit_reset(ctx);
+static int
+sc_mark_corrupt(struct scrub_ctx *ctx)
+{
+	return vos_iter_process(ctx->sc_vos_iter_handle,
+				VOS_ITER_PROC_OP_MARK_CORRUPT, NULL);
 }
 
-/*
- * DAOS_CSUM_SCRUB_DISABLED_WAIT_SEC can be set in the server config to change
- * how long to wait before checking again if the scrubber is enabled.
- */
-static uint32_t
-seconds_to_wait_while_disabled()
+static int
+sc_pool_drain(struct scrub_ctx *ctx)
 {
-	char *sec = getenv("DAOS_CSUM_SCRUB_DISABLED_WAIT_SEC");
+	D_ASSERT(ctx->sc_drain_pool_tgt_fn);
 
-	return sec != NULL ? atoll(sec) : 5;
+	return ctx->sc_drain_pool_tgt_fn(ctx->sc_pool);
 }
 
-void
-sc_yield_or_sleep(struct scrub_ctx *ctx)
+static bool
+sc_should_evict(struct scrub_ctx *ctx)
 {
-	struct timespec	now, diff;
-	uint32_t	left_sec;
 
-	d_gettime(&now);
-	diff = d_timediff(ctx->sc_pool_start_scrub, now);
-
-	if (diff.tv_sec < sc_freq(ctx)) {
-		left_sec = sc_freq(ctx) - diff.tv_sec;
-		sc_sleep(ctx, left_sec * 1000);
-	} else {
-		sc_yield(ctx);
-	}
+	return sc_thresh(ctx) > 0 && /* threshold set */
+	       ctx->sc_pool_tgt_corrupted_detected >= /* hit or exceeded */
+	       sc_thresh(ctx);
 }
 
-static inline bool
-sc_scrub_enabled(struct scrub_ctx *ctx)
+static bool
+sc_is_nvme(struct scrub_ctx *ctx)
 {
-	return sc_schedule(ctx) != DAOS_SCRUB_SCHED_OFF && sc_freq(ctx) > 0;
+	return bio_iov2media(ctx->sc_cur_biov) == DAOS_MEDIA_NVME;
 }
 
-static void
-sc_verify_finish(struct scrub_ctx *ctx)
+static int
+sc_handle_corruption(struct scrub_ctx *ctx)
 {
-	sc_csum_calc_inc(ctx);
-	sc_m_pool_csum_inc(ctx);
-	sc_yield_sleep_while_running(ctx);
+	int rc;
+
+	sc_raise_ras(ctx);
+	sc_m_pool_corr_inc(ctx);
+	rc = sc_mark_corrupt(ctx);
+
+	if (sc_is_nvme(ctx))
+		bio_log_csum_err(ctx->sc_dmi->dmi_nvme_ctxt);
+	if (rc != 0) {
+		/* Log error but don't let it stop the scrubbing process */
+		D_ERROR("Error trying to mark corrupt: "DF_RC"\n", DP_RC(rc));
+		rc = 0;
+	}
+	ctx->sc_pool_tgt_corrupted_detected++;
+	D_ERROR("Checksum scrubber found corruption. %d so far.\n",
+		ctx->sc_pool_tgt_corrupted_detected);
+	if (sc_should_evict(ctx)) {
+		D_ERROR("Corruption threshold reached. %d >= %d",
+			ctx->sc_pool_tgt_corrupted_detected,
+			sc_thresh(ctx));
+		d_tm_set_counter(ctx->sc_metrics.scm_csum_calcs, 0);
+		d_tm_set_counter(ctx->sc_metrics.scm_csum_calcs_last, 0);
+		rc = sc_pool_drain(ctx);
+		if (rc != 0)
+			D_ERROR("Drain error: "DF_RC"\n", DP_RC(rc));
+
+		rc = -DER_SHUTDOWN;
```

Comment:
This is to inform scrubbing ULT to exit? I think it could be improved to ensure above drain operation succeeds (the target is in DRAIN state) before exits.

Reply:
Correct. I think I'll leave it like this for now since the scrubber's auto eviction feature may go away in the near future and use the SSD auto eviction feature.